src/NHibernate/Type/GenericMapType.cs

```diff
@@ -116,5 +117,12 @@ public override object IndexOf(object collection, object element)
 				.Select(pair => pair.Key)
 				.FirstOrDefault();
 		}
+
+		protected override bool AreCollectionElementsEqual(IEnumerable original, IEnumerable target)
+		{
+			var first = (IDictionary<TKey, TValue>)original;
+			var second = (IDictionary<TKey, TValue>)target;
+			return first.Count == second.Count && first.All(keyValue => second.Contains(keyValue));
```

Comment:
Hm. In what cases do you mean that is needed? Maybe I'm too tired now and miss something obvious but... 
I mean KeyValuePair is a struct and will check its data for equality by default?